### PR TITLE
Shared class cache updates for kernel and application classloader

### PR DESCRIPTION
**Pull request by jhanders34**
Updates improve how shared class cache url is calculated for all the different application artifact options.  The improvements should also help loose applications which probably didn't have shared class caching happening, but now will.  

Additionally there are other minor changes related to changes I found while doing these code changes for AppClassLoader and looking at heap dump to find other way to reduce heap usage.

These changes help resolve the problem in issue #6042.  Previously a PR was put in to work around the problem, but this PR fixes it for real.

### CODE REVIEW

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#build

File: dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainer.java [hunk 1]
```diff
@@ -1167,13 +1167,36 @@ public ZipFileEntry getEntry(String entryPath, boolean normalized) {
         }
     }
 
+    @Trivial 
+    URL createSharedClassCacheURL(String r_entryPath) {
+        File useArchiveFile = getArchiveFile();
```
Comment (andymc12) [inline dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainer.java]: `useArchiveFile` sounds more like a boolean (i.e. used to determine if we are using an archive file or not).  Would `archiveFile` be a better variable name?

File: dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainerUtils.java [hunk 2]
```diff
@@ -453,36 +454,170 @@ public int getOffset() {
...
         }
         
         @Trivial
+        @Override
         public long getTime() {
             return time;
         }
     }
     
+    private static class DirZipEntryData implements ZipEntryData {
```
Comment (andymc12) [inline dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainerUtils.java]: I'm wondering if you could save some code by using inheritance here...  not a show-stopper though.

File: dev/com.ibm.ws.artifact/src/com/ibm/wsspi/artifact/ArtifactEntry.java [hunk 3]
```diff
@@ -79,12 +79,23 @@
...
 
-    //Concepts absent, but under discussion. 
+    /**
+     * Returns the URL to use for the shared class cache to track if the class changes.
+     *
+     * For a jar it is the path of the jar. If it is another archive type, it is the patch to the archive with
+     * the entry, i.e. jar:file:<file>~/package/className.class. For a directory it is the directory URL.
+     *
+     * @param classFileResourceName class name in relative path form a/b/c.class
+     * @return URL to be used for the IBM / OpenJ9 shared class cache function.
```
Comment (andymc12) [inline dev/com.ibm.ws.artifact/src/com/ibm/wsspi/artifact/ArtifactEntry.java]: Like before - might want to mention that this could return null.

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/PooledWsByteBufferImpl.java [hunk 4]
```diff
@@ -44,7 +44,7 @@
     protected int actionState = COPY_ALL_INIT;
     protected final Object actionAccess = new Object();
 
-    private final Object identifier;
+    private final int identifier;
```
Comment (andymc12) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/PooledWsByteBufferImpl.java]: Are these byte buffer changes related to the shared class cache changes?  Maybe they should go in a separate PR?

File: dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainer.java (hunk 1 above)
Comment (tbitonti) [reply dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainer.java]: 'archiveFile' is already used as an instance variable name.  'use' is added to prevent a compiler warning.
I've been putting 'use' as a prefix, but other values, say, 'my' or 'a' would work just as well.

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/PooledWsByteBufferImpl.java (hunk 4 above)
Comment (jhanders34) [reply dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/PooledWsByteBufferImpl.java]: I update the PR comment to include information that I included some other things found along the way.  Hopefully that satisfies this review comment.

File: dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainer.java (hunk 1 above)
Comment (jhanders34) [reply dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainer.java]: This code is now gone so it is a moot point.

File: dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainerUtils.java (hunk 2 above)
Comment (jhanders34) [reply dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainerUtils.java]: I added some inheritance to address your review comments.

File: dev/com.ibm.ws.artifact/src/com/ibm/wsspi/artifact/ArtifactEntry.java (hunk 3 above)
Comment (jhanders34) [reply dev/com.ibm.ws.artifact/src/com/ibm/wsspi/artifact/ArtifactEntry.java]: As before the new method does mention it returning null.

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/PooledWsByteBufferImpl.java (hunk 4 above)
Comment (jhanders34) [reply dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/PooledWsByteBufferImpl.java]: Update the PR comment to state that fixing some other things found along the way.

File: dev/com.ibm.ws.adaptable.module/src/com/ibm/wsspi/adaptable/module/Entry.java [hunk 5]
```diff
@@ -88,17 +88,27 @@
...
     public Container getRoot();
 
+    /**
+     * Returns the URL to use for the shared class cache to track if the class changes.
+     *
+     * For a jar it is the path of the jar. If it is another archive type, it is the patch to the archive with
+     * the entry, i.e. jar:file:<file>~/package/className.class. For a directory it is the directory URL.
+     *
+     * @param classFileResourceName class name in relative path form a/b/c.class
+     * @return URL to be used for the IBM / OpenJ9 shared class cache function.
```
[the remark this answers is not in the document]
Comment (jhanders34) [reply dev/com.ibm.ws.adaptable.module/src/com/ibm/wsspi/adaptable/module/Entry.java]: The new method indicates that it can return null.

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/WsByteBufferImpl.java [hunk 6]
```diff
@@ -73,8 +73,8 @@
     protected int quickBufferAction = NOT_ACTIVATED;
 
     // min/max are inclusive, so threshold will be this number + 1
-    private int GET_THRESHOLD = 1023; // minimum amount to copy from Direct to NonDirect
-    private int PUT_THRESHOLD = 2047; // maximum number of bytes between non-contiguous puts that will be bridged.
+    private final int GET_THRESHOLD = 1023; // minimum amount to copy from Direct to NonDirect
```
Comment (vijaysun-omr) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/WsByteBufferImpl.java]: Making these static (in addition to final) could result in better JIT optimization (folding away of the loads from memory to be constants instead).

File: dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainerUtils.java [hunk 7]
```diff
@@ -406,83 +406,139 @@ private static boolean isChildOf(
...
+
+        abstract boolean isDirectory();
+
+        abstract int getSize();
+    }
 
-        private final String path;
-        private final String r_path;
+    private static class FileZipEntryData extends ZipEntryData {
+        @Trivial
```
[the remark this answers is not in the document]
Comment (jhanders34) [reply dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileContainerUtils.java]: Updated accordingly.

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/WsByteBufferImpl.java (hunk 6 above)
Comment (jhanders34) [reply dev/com.ibm.ws.channelfw/src/com/ibm/ws/bytebuffer/internal/WsByteBufferImpl.java]: Eclipse added the final automatically as part of the automatic class formatting.  It was not an explicit change to the code.  I have updated the code now though to use static.

**Comment by jhanders34** [comment]
#build

**Comment by wraschke** [comment]
This PR causes https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.workitem.viewWorkItem&id=261136
